nomos-services/mempool/src/tx/service.rs

```diff
@@ -6,117 +6,157 @@ pub mod openapi {
+    NetworkAdapter: NetworkAdapterTrait,
+    NetworkAdapter::Settings: Clone,
+    RecoveryBackend: RecoveryBackendTrait,
 {
     const SERVICE_ID: ServiceId = "mempool-cl";
-    type Settings = TxMempoolSettings<P::Settings, N::Settings>;
-    type State = NoState<Self::Settings>;
-    type StateOperator = NoOperator<Self::State, Self::Settings>;
-    type Message = MempoolMsg<
-        <P as MemPool>::BlockId,
-        <P as MemPool>::Item,
-        <P as MemPool>::Item,
-        <P as MemPool>::Key,
-    >;
+    type Settings = TxMempoolSettings<Pool::Settings, NetworkAdapter::Settings>;
+    type State = TxMempoolState<Pool::RecoveryState, Pool::Settings, NetworkAdapter::Settings>;
+    type StateOperator = RecoveryOperator<RecoveryBackend>;
+    type Message = MempoolMsg<Pool::BlockId, Pool::Item, Pool::Item, Pool::Key>;
 }
 
 #[async_trait::async_trait]
-impl<N, P> ServiceCore for TxMempoolService<N, P>
+impl<Pool, NetworkAdapter, RecoveryBackend> ServiceCore
+    for GenericTxMempoolService<Pool, NetworkAdapter, RecoveryBackend>
 where
-    P: MemPool + Send + 'static,
-    P::Settings: Clone + Send + Sync + 'static,
-    N::Settings: Clone + Send + Sync + 'static,
-    P::Item: Clone + Debug + Send + Sync + 'static,
-    P::Key: Debug + Send + Sync + 'static,
-    P::BlockId: Send + Debug + 'static,
-    N: NetworkAdapter<Payload = P::Item, Key = P::Key> + Send + Sync + 'static,
+    Pool: RecoverableMempool + Send,
+    Pool::RecoveryState: Debug + Send + Sync,
+    Pool::Key: Send,
+    Pool::Item: Clone + Send + 'static,
+    Pool::BlockId: Send,
+    Pool::Settings: Clone + Sync + Send,
+    NetworkAdapter: NetworkAdapterTrait<Payload = Pool::Item, Key = Pool::Key> + Send,
+    NetworkAdapter::Settings: Clone + Send + Sync + 'static,
+    RecoveryBackend: RecoveryBackendTrait + Send,
 {
     fn init(
-        service_state: OpaqueServiceStateHandle<Self>,
-        _init_state: Self::State,
+        service_state: ServiceStateHandle<Self::Message, Self::Settings, Self::State>,
+        init_state: Self::State,
     ) -> Result<Self, overwatch_rs::DynError> {
-        let network_relay = service_state.overwatch_handle.relay();
+        tracing::trace!(
+            "Initializing TxMempoolService with initial state {:#?}",
+            init_state.pool
+        );
         let settings = service_state.settings_reader.get_updated_settings();
+        let recovered_pool = init_state.pool.map_or_else(
+            || Pool::new(settings.pool.clone()),
+            |recovered_pool| Pool::recover(settings.pool.clone(), recovered_pool),
+        );
 
-        Ok(Self {
-            service_state,
-            network_relay,
-            pool: P::new(settings.backend),
-        })
+        Ok(Self::new(recovered_pool, service_state))
     }
 
     async fn run(mut self) -> Result<(), overwatch_rs::DynError> {
-        let Self {
-            mut service_state,
-            network_relay,
-            mut pool,
-            ..
-        } = self;
```

Comment:
I'm also curious, is there a lint for this as well?

Reply:
No need to change. I was curious as well. But I would prefer that if it is just personal preference, we keep what is there unless there is a specific reason for that. I think in this case, it is ok as we changed some `async` methods to `sync` so borrowing of `self` is not affected. But in most services, you will end up having problems as you need different parts of `self` in different async contexts.